From 997a5babb3a0b9a31ffe07356c7d06e0ec730797 Mon Sep 17 00:00:00 2001
From: smomin7 <smomin@vicon-security.com>
Date: Wed, 2 Apr 2025 21:37:09 +0530
Subject: [PATCH 14/18] gstidentity:  changes for zoom.

---
 subprojects/gstreamer/plugins/elements/gstidentity.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/subprojects/gstreamer/plugins/elements/gstidentity.c b/subprojects/gstreamer/plugins/elements/gstidentity.c
index e3063b9327..d0ec53ae81 100644
--- a/subprojects/gstreamer/plugins/elements/gstidentity.c
+++ b/subprojects/gstreamer/plugins/elements/gstidentity.c
@@ -75,6 +75,7 @@ enum
 #define DEFAULT_CHECK_IMPERFECT_OFFSET    FALSE
 #define DEFAULT_SIGNAL_HANDOFFS           TRUE
 #define DEFAULT_TS_OFFSET               0
+#define DEFAULT_LAST_BUFFER_TIMESTAMP   0
 #define DEFAULT_DROP_ALLOCATION         FALSE
 #define DEFAULT_EOS_AFTER               -1
 
@@ -92,6 +93,7 @@ enum
   PROP_DUMP,
   PROP_SYNC,
   PROP_TS_OFFSET,
+  PROP_LAST_BUFFER_TIMESTAMP,
   PROP_CHECK_IMPERFECT_TIMESTAMP,
   PROP_CHECK_IMPERFECT_OFFSET,
   PROP_SIGNAL_HANDOFFS,
@@ -216,6 +218,11 @@ gst_identity_class_init (GstIdentityClass * klass)
           "Timestamp offset in nanoseconds for synchronisation, negative for earlier sync",
           G_MININT64, G_MAXINT64, DEFAULT_TS_OFFSET,
           G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+  g_object_class_install_property (gobject_class, PROP_LAST_BUFFER_TIMESTAMP,
+      g_param_spec_int64 ("last-buffer-timestamp", "last buffer timestamp",
+          "Timestamp offset in nanoseconds for synchronisation, negative for earlier sync",
+          0, G_MAXINT64, DEFAULT_LAST_BUFFER_TIMESTAMP,
+          G_PARAM_READABLE | G_PARAM_STATIC_STRINGS));
   g_object_class_install_property (gobject_class,
       PROP_CHECK_IMPERFECT_TIMESTAMP,
       g_param_spec_boolean ("check-imperfect-timestamp",
@@ -1016,6 +1023,9 @@ gst_identity_get_property (GObject * object, guint prop_id, GValue * value,
     case PROP_TS_OFFSET:
       g_value_set_int64 (value, identity->ts_offset);
       break;
+    case PROP_LAST_BUFFER_TIMESTAMP:
+      g_value_set_int64 (value, identity->prev_timestamp);
+      break;
     case PROP_CHECK_IMPERFECT_TIMESTAMP:
       g_value_set_boolean (value, identity->check_imperfect_timestamp);
       break;
-- 
2.40.1.windows.1

